File: services/api/app/schemas/validation.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field

from app.schemas.common import JsonDict, StrictModel, ValidationStatus

Severity = Literal["info", "low", "medium", "high", "critical"]
CheckStatus = Literal["passed", "failed", "skipped"]
FileChangeStatus = Literal["added", "modified", "deleted", "renamed"]
DependencyChangeAction = Literal["added", "updated", "removed"]
ValidationMode = Literal["bundle", "patch", "repository", "dry-run"]


class ChangedFile(StrictModel):
    path: str = Field(min_length=1)
    status: FileChangeStatus = "modified"
    digest: str | None = None
    previous_path: str | None = None


class DependencyChange(StrictModel):
    ecosystem: str = Field(min_length=1)
    name: str = Field(min_length=1)
    action: DependencyChangeAction = "added"
    version: str | None = None
    approved: bool = False
    reason: str | None = None


class ValidationArtifact(StrictModel):
    kind: str = Field(min_length=1)
    path: str = Field(min_length=1)
    digest: str | None = None
    verified: bool = False


class ValidationRequest(StrictModel):
    """A generated repo or patch submitted for Matrix contract validation.

    This accepts references instead of raw archives so the API can support browser uploads,
    object-storage uploads, GitHub PR refs, or local dev fixtures without changing the contract.
    """

    schema_version: str = "matrix.builder.validation-request/v1"
    bundle_id: str | None = None
    mode: ValidationMode = "patch"
    changed_files: list[ChangedFile] = Field(default_factory=list)
    dependency_changes: list[DependencyChange] = Field(default_factory=list)
    artifacts: list[ValidationArtifact] = Field(default_factory=list)
    standards_lock_digest: str | None = None
    matrix_blueprint_digest: str | None = None
    repository_archive_ref: str | None = None
    patch_ref: str | None = None
    metadata: JsonDict = Field(default_factory=dict)


class ValidationViolation(StrictModel):
    rule_id: str
    severity: Severity
    message: str
    path: str | None = None
    remediation: str | None = None


class ValidationCheck(StrictModel):
    check_id: str
    status: CheckStatus
    message: str | None = None


class ValidationReport(StrictModel):
    report_id: str = "validation_not_run"
    bundle_id: str = "bundle_demo"
    status: ValidationStatus = ValidationStatus.NOT_RUN
    score: int = Field(ge=0, le=100)
    violations: list[ValidationViolation] = Field(default_factory=list)
    repair_prompt: str | None = None
    checks: list[ValidationCheck] = Field(default_factory=list)
    approved: bool = False
    matrixhub_publishable: bool = False
    summary: str | None = None
    created_at: datetime | None = None

File: apps/web/src/types/validation.ts
export type ValidationStatus = "not-run" | "approved" | "needs-repair" | "rejected";
export type ViolationSeverity = "info" | "low" | "medium" | "high" | "critical";

export interface ValidationViolation {
  rule_id: string;
  severity: ViolationSeverity;
  path?: string | null;
  message: string;
  remediation?: string | null;
}

export interface ValidationCheck {
  check_id: string;
  status: "passed" | "failed" | "skipped";
  message?: string | null;
}

export interface ValidationReport {
  report_id: string;
  bundle_id: string;
  status: ValidationStatus;
  score: number;
  violations: ValidationViolation[];
  repair_prompt?: string | null;
  checks: ValidationCheck[];
  approved: boolean;
  matrixhub_publishable: boolean;
  summary?: string | null;
  created_at?: string | null;
}

export const demoValidationReports: ValidationReport[] = [
  {
    report_id: "val_approved_demo",
    bundle_id: "bundle_demo_standard",
    status: "approved",
    score: 100,
    violations: [],
    checks: [
      { check_id: "forbidden_file_changes_absent", status: "passed", message: "No Matrix control files changed." },
      { check_id: "dependency_drift_absent", status: "passed", message: "No unapproved dependency changes." }
    ],
    approved: true,
    matrixhub_publishable: true,
    summary: "Approved: the AI-coder output stayed inside the Matrix contract."
  },
  {
    report_id: "val_repair_demo",
    bundle_id: "bundle_demo_standard",
    status: "needs-repair",
    score: 80,
    violations: [
      {
        rule_id: "RMD-105",
        severity: "high",
        path: "npm:left-pad",
        message: "Unapproved dependency added.",
        remediation: "Remove the dependency or request explicit approval."
      }
    ],
    repair_prompt: "Repair only the dependency drift. Do not change architecture or Matrix control files.",
    checks: [{ check_id: "dependency_drift_absent", status: "failed", message: "npm:left-pad" }],
    approved: false,
    matrixhub_publishable: false,
    summary: "Needs repair: dependency drift must be fixed before approval."
  },
  {
    report_id: "val_rejected_demo",
    bundle_id: "bundle_demo_standard",
    status: "rejected",
    score: 30,
    violations: [
      {
        rule_id: "RMD-103",
        severity: "critical",
        path: "MATRIX_BLUEPRINT.yaml",
        message: "Immutable Matrix blueprint was modified.",
        remediation: "Restore the file from the original Matrix Bundle."
      }
    ],
    repair_prompt: "Restore MATRIX_BLUEPRINT.yaml from the Matrix Bundle. AI coders are workers, not architects.",
    checks: [{ check_id: "forbidden_file_changes_absent", status: "failed", message: "MATRIX_BLUEPRINT.yaml" }],
    approved: false,
    matrixhub_publishable: false,
    summary: "Rejected: critical Matrix contract drift was detected."
  }
];
